### ISSUE TITLE
Consider removing ruleMessageId and updating rule message lookup semantics

### ISSUE THREAD

**Issue by michaelcfanning** [oasis-tcs/sarif-spec#216]
Why do we still have ruleMessageId? A message object contains a messageId already, which is sufficient information to locate it. Note that even with a ruleMessageId, the message instance is likely required in order to persist the arguments for the ruleMessage (assuming the message is a format string, of course).

I think we can eliminate the ruleMessageId property. Now we need to clarify the message string look-up semantics for results:

in the presence of a non-null message.messageId, consult resources.rules by rule id, to attempt to locate the string. if not present, fall back to resources for look-up.

I think it may be an oversight that ruleMessageId is still hanging around, as we don't see a corresponding ruleRichMessageId. We left this behind after we did some useful reworking of the message class...

**Comment by lgolding** [comment]
TL;DR: Good catch! You are right, we can dispense with `result.ruleMessageId`.

I left `ruleMessageId` in place intentionally; it wasn't an oversight. I had defined the semantics of `message.messageId` to be "look this string up in `resources.messageStrings`" (see §3.9.6.2, "Message string lookup procedure".) But rule messages don't live in `resources.messageStrings`, they live in `resources.rules[ruleId].messageStrings`. So, I reasoned, I needed a different property.

I didn't leave that property behind because I thought it would allow you to dispense with the `message` object -- as you say, you still need the `message` object to hold the arguments. I left it behind because I wanted a property whose semantics were "Look up this string in `resources.rules[ruleId].messageStrings`," and I didn't want to put such a `result`-specific property in the generally useful `message` object.

But your proposal -- to look in `resources.rules` first (in the case of a `result` message), and then to fall back to `resources.messageStrings` -- solves the problem.

Now, not having a corresponding `result.ruleRichMessageId` -- that _was_ an oversight. But it's moot now, with your proposal.

**Comment by lgolding** [comment]
@michaelcfanning Yes.
